File: csharp/lib/Property/ChannelInfoProperty.cs
/*
 * SPDX-FileCopyrightText: 2017-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using System.Collections.Generic;
using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Property for channel information.
    /// </summary>
    [MessagePackObject]
    public class ChannelInfoProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "channel_info_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// Channel information list (Key=Channel ID).
        /// </summary>
        [Key("channels")]
        public Dictionary<long, ChannelInfo> Channels { get; set; } =
            new Dictionary<long, ChannelInfo>();

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return $"{this.Key}({string.Join(", ", this.Channels)})";
        }
    }

    /// <summary>
    /// Channel information.
    /// </summary>
    [MessagePackObject]
    public class ChannelInfo
    {
        /// <summary>
        /// Type of raw data.
        /// </summary>
        [Key("raw_data_type")]
        public string RawDataType { get; set; } = string.Empty;

        /// <summary>
        /// Channel description.
        /// </summary>
        [Key("description")]
        public string Description { get; set; } = string.Empty;

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return $"{this.RawDataType}: {this.Description}";
        }
    }
}

File: java/senscord/src/senscord_properties/ChannelInfoProperty.java
/*
 * SPDX-FileCopyrightText: 2019 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_properties;

import com.sun.jna.Structure;
import java.util.Arrays;
import java.util.List;
import senscord_types.ChannelInfo;

/**
 * Property for channel information.
 * 
 * @see "senscord::ChannelInfoProperty"
 */
public class ChannelInfoProperty extends Structure {
    public int count;
    public ChannelInfo[] channel_info;

    public static class ByReference extends ChannelInfoProperty implements Structure.ByReference {
    }

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("count", "channel_info");
    }

    public ChannelInfoProperty() {
        channel_info = new ChannelInfo[128];
    }

    /**
     * Get channel info count.
     * 
     * @return channel info count
     */
    public int GetCount() {
        return count;
    }

    /**
     * Get channel info class array.
     * 
     * @return channel info class array
     */
    public ChannelInfo[] GetChannelInfo() {
        return channel_info;
    }

}
